---
title: "Haplotype Network SQR"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# How-to Generate a Phased Haplotype Network from a VCF file 

The following tutorial describes the steps to generate a haplotype network from a VCF file. The following example uses an unphased VCF file containing variant information from 20 individuals per population for 10 populations, which can be dowloaded here (ADD HYPERLINK TO GITHUB FILE LOCATION). The general workflow is as followed 

1) Split the multi-population VCF into a VCF for each population 
2) Phase the population VCF
3) Split VCF by individual 
4) Generate consensus fasta files for each haplotype for each individual 
5) Rename fasta headers and concatenate fastas
6) Generate network
7) Generate metadata/annotation files
8) Visualize and improve  

This tutorial requires the following programs and was run using the following versions:
* VCFtools v.0.1.16
* BCFtools v.1.10.2 (using htslib v.1.10.2)
* Popgen Pipeline Platform (PPP) v.0.1.12
* Python v.3.7.7
* Beagle v.5.1
* Mafft v.7.429
* R v.4.1.2
* Population and Evolutionary Genetics Analysis System (‘pegas’) v.1.1
* Analysis of Phylogenetic and Evolution (‘ape’) v.5.6.2
* Cytoscape v.3.9.1


## Haplotype Network

```{r packages}
# Read in packages 
library("ape")
library("pegas")
library('ggplot2')
library(network)
library(dplyr)
library(tidyr)
```


Haplotypes will be determined using pegas.
Once generated, the data will be pulled into cytoscape for better visualization and annotations.
For cytoscape, you need a network table, a table that contains the edge interactions with a column that contains step size, and a table that has information linking the haplotypes to population data. These will be generated below, but can be generated using a variety of methods using other R functions such as igraph and network. 

Example of what necessary network tables look like
This is technically just an edge list!
| Node_1 | Node_2 |
|--------|--------|
| 1      | 3      |
| 1      | 31     |
| 2      | 3      |
| 4      | 7      |
| 5      | 23     |
| 6      | 24     |
| 6      | 69     |
| 6      | 75     |
| 7      | 16     |
| 7      | 19     |
| 7      | 20     |
| 7      | 43     |
| 7      | 66     |
| 7      | 74     |


Edge info that contains the number of mutations between nodes (haplotypes)
| **Shared name**        | **step** |
|------------------------|----------|
| 1 (interacts with) 31  | 1        |
| 2 (interacts with) 3   | 1        |
| 4 (interacts with) 7   | 1        |
| 7 (interacts with) 19  | 1        |
| 7 (interacts with) 20  | 1        |
| 7 (interacts with) 43  | 1        |
| 9 (interacts with) 14  | 1        |
| 9 (interacts with) 44  | 1        |
| 17 (interacts with) 29 | 1        |
| 31 (interacts with) 32 | 1        |
| 35 (interacts with) 77 | 1        |

Population information about each haplotype (in wide format)
| Haplotype | Banos | Bonita | Esperanza | Gloria | Ixta |
|-----------|-------|--------|-----------|--------|------|
| 1         | 26    | 0      | 0         | 1      | 0    |
| 2         | 9     | 0      | 0         | 0      | 0    |
| 3         | 5     | 0      | 0         | 0      | 0    |
| 4         | 0     | 3      | 0         | 0      | 0    |
| 5         | 0     | 1      | 0         | 0      | 0    |
| 6         | 0     | 1      | 0         | 0      | 0    |
| 7         | 0     | 4      | 0         | 0      | 0    |
| 8         | 0     | 1      | 0         | 0      | 0    |


# ETHE1
## ETHE1.a
This haplotype network is for LOC106917689 which we refer to as ETHE1.a in this manuscript 
Located on NW_015094773.1 (309149..314096, complement)

The fasta file was phased using beagle and then a consensus seq was generated for each haplotype for each individual
The fasta file contains 398 haplotype prior to filtering 
```{r generate ethe1 network}
# read in data in fasta format
data<- read.dna("/Users/kararyan/Haplotype_Network/merged_ethe1a.fasta", format = "fasta")
data

# determine the haplotypes
haps = haplotype(data)
haps
# generate the network
Net <- haploNet(haps)

# convert network of class HaploNet to more generalize network object 
# this is an undirected network - default is directed so you must set this option
Net.ethe1 <- as.network(Net, directed = FALSE)

# This conversion loosing info about edge attributes 
# We can generate this after
Net.ethe1

# Get the edge list out of the network 
Net.ethe1.edgelist <- as.edgelist(Net.ethe1)
# convert edge list to dataframe
ethe1.network<- as.data.frame(Net.ethe1.edgelist)
# write table to be visualized in cytoscape
write.csv(Net.ethe1.edgelist, file = "ethe1.network.csv", quotes = FALSE, row.names = FALSE)
```
Next, we need to collect useful information about node size (how many individuals have each haplotype) and the population makeup of each node. Additionally, this step will gather information about node size  
```{r ethe1 net stats}
# determine distance between haplotypes (numbe of mutations) 
# this outputs a pairwise dist matrix. 
d <- dist.dna(haps, "N")

# add distance info ad edge links
nt <- rmst(d, quiet = TRUE)

# summarize the haplotypes to get totals (node sizes) and add labels 
sz <- summary(haps)
nt.labs <- attr(nt, "labels")
sz <- sz[nt.labs]

# generate table that connects each individual to their haplotypes (indv as cols, haps as rows)
ind.hap<-with(stack(setNames(attr(haps, 'index'), rownames(haps))), 
table(hap=ind, individuals=rownames(data)[values]))

# convert to a data frame
mydata <- as.data.frame(ind.hap)

# remove extra columns where there is no haplotype
good <- mydata[mydata$Freq == 1,]

# Split the indv names to remove the _H1 and _H2 that extists in labels 
indv <- strsplit(as.character(good$individuals), '_')

# split to get location by removing sample number 
locations <- strsplit(as.character(good$individuals), "(?=[A-Za-z])(?<=[0-9])|(?=[0-9])(?<=[A-Za-z])", perl=TRUE)
locations <- sapply(locations, "[[", 1)

# add location information to the table of hapltypes 
new.hap<- table(good$hap, locations)
# convert to data frame
new.hap <- as.data.frame(new.hap)
#rename the first column that contains haplotype info 
new.hap<-rename(new.hap, "Haplotype" = "Var1")

# convert roman numerals to numbers
new.hap$Haplotype <- as.numeric(new.hap$Haplotype)

# pivot the table to wide for annotations in cytoscape and save the table
new.hap.wide <- pivot_wider(new.hap, names_from = "locations", values_from = "Freq")
write.csv(new.hap.wide, "ethe1.hap.info.wide.csv", quote = FALSE, row.names = FALSE)

# set step info table
# I could not convert the haploNet object to a datatable and retain step info
# As a workaround, I save the network then read it in as a data frame
write.csv(Net, "ethe1.step.info.csv", quote = FALSE, row.names = FALSE)
step_size <- read.csv("ethe1.step.info.csv")

# set edge interactions so cytoscape can read 
step_size$`shared name` <- paste0(step_size$X.1, " (interacts with) ", step_size$X)

# only select the interaction and step size table
step_size %>%
  select(`shared name`, step) -> step_size
# write step size table 
write.csv(step_size, "ethe1.step.info.csv", quote = FALSE, row.names = FALSE)
```
This end the basic tutorial for generate necesary files for haplotype network visualiztion. Below is filtering options and graphing for smaller number of nodes. This can still be comp intensive in R and should be run on kamiak! 



## ETHE1 singletons removed
```{r read in fasta file}
no_sing_haps <- subset(haps, minfreq = 2)
no_sing_haps
Net_1 <- haploNet(no_sing_haps)
d <- dist.dna(no_sing_haps, "N")
nt <- rmst(d, quiet = TRUE)
sz <- summary(no_sing_haps)
nt.labs <- attr(nt, "labels")
sz <- sz[nt.labs]
plot(Net_1, size = sz)
# convert network of class HaploNet to more generalize network object 
# this is an undirected network - default is directed so you must set this option
Net.ethe1 <- as.network(Net_1, directed = FALSE)
# This conversion loosing info about edge attributes 
# We can generate this after
Net.ethe1

# Get the edge list out of the network 
Net.ethe1.edgelist <- as.edgelist(Net.ethe1)
# convert edge list to dataframe
ethe1.network<- as.data.frame(Net.ethe1.edgelist)
# write table to be visualized in cytoscape
write.csv(Net.ethe1.edgelist, file = "ethe1_rm1.network.csv", row.names = FALSE)
```

```{r}
# generate table that connects each individual to their haplotypes (indv as cols, haps as rows)
ind.hap<-with(stack(setNames(attr(no_sing_haps, 'index'), rownames(no_sing_haps))),table(hap=ind, individuals=rownames(data)[values]))
# convert to a data frame
mydata <- as.data.frame(ind.hap)
# remove extra columns where there is no haplotype
good <- mydata[mydata$Freq == 1,]
# Split the indv names to remove the _H1 and _H2 that extists in labels 
indv <- strsplit(as.character(good$individuals), '_')
# split to get location by removing sample number 
locations <- strsplit(as.character(good$individuals), "(?=[A-Za-z])(?<=[0-9])|(?=[0-9])(?<=[A-Za-z])", perl=TRUE)
locations <- sapply(locations, "[[", 1)
# add location information to the table of haplotypes 
new.hap<- table(good$hap, locations)
# convert to data frame
new.hap <- as.data.frame(new.hap)
#rename the first column that contains haplotype info 
new.hap<-rename(new.hap, "Haplotype" = "Var1")
# convert roman numerals to numbers
new.hap$Haplotype <- as.numeric(new.hap$Haplotype)

# pivot the table to wide for annotations in cytoscape and save the table
new.hap.wide <- pivot_wider(new.hap, names_from = "locations", values_from = "Freq")
write.csv(new.hap.wide, "ethe1_rm1.hap.info.wide.csv", quote = FALSE, row.names = FALSE)

# set step info table
# I could not convert the haploNet object to a datatable and retain step info
# As a workaround, I save the network then read it in as a data frame
write.csv(Net_1, "ethe1_rm1.step.info.csv", quote = FALSE, row.names = FALSE)
step_size <- read.csv("ethe1_rm1.step.info.csv")

# set edge interactions so cytoscape can read 
step_size$`shared name` <- paste0(step_size$X.1, " (interacts with) ", step_size$X)

# only select the interaction and step size table
step_size %>%
  select(`shared name`, step) -> step_size
# write step size table 
write.csv(step_size, "ethe1_rm1.step.info.csv", quote = FALSE, row.names = FALSE)
```















```{r }
pdf("ethe1.a_rm1.pdf")
plot(Net_1, size=attr(Net_1, "freq"), scale.ratio = 2, cex = 0.8, pie=new.hap, fast = FALSE)
legend('topleft', colnames(new.hap), col=rainbow(ncol(new.hap)), pch=20, cex=1)
dev.off()
```


```{r ethe1 doubletons removed}
no_doub_haps <- subset(haps, minfreq = 3 )
no_doub_haps 
Net_2 = haploNet(no_doub_haps)
d <- dist.dna(no_doub_haps, "N")
nt <- rmst(d, quiet = TRUE)
nt
sz <- summary(no_doub_haps)
nt.labs <- attr(nt, "labels")
sz <- sz[nt.labs]
plot(Net_2, size = sz)
ind.hap<-with(stack(setNames(attr(no_doub_haps, 'index'), rownames(no_doub_haps))), 
table(hap=ind, individuals=rownames(data)[values]))
ind.hap
mydata <- as.data.frame(ind.hap)

good <- mydata[mydata$Freq == 1,]
good
indv <- strsplit(as.character(good$individuals), '_')
locations <- strsplit(as.character(good$individuals), "(?=[A-Za-z])(?<=[0-9])|(?=[0-9])(?<=[A-Za-z])", perl=TRUE)
locations <- sapply(locations, "[[", 1)
new.hap<- table(good$hap, locations)
summary(new.hap)

colors = c('#EEE8AA', '#CCFFFF', '#FFFF33', '#FFDB58', '#98AFC7', '#EAC117', '#FBB117', '#151B54', '#1F45FC' ,'#00BFFF' )


# pdf("ethe1.a_rm2.pdf")
plot(Net_2, size=attr(Net_2, "freq"), scale.ratio = 2, cex = 0.8, pie=new.hap, fast = FALSE, legend = c(-100,150) )
#legend('topleft', colnames(new.hap), col=rainbow(ncol(new.hap)), pch=20, cex=1)

#dev.off()

rainbow(ncol(new.hap))
legend = c(-100,150)
```



# SQOR

```{r read in fasta file}
data<- read.dna("/Users/kararyan/Haplotype_Network/sqor_merged_aligned_haplotypes.fasta", format = "fasta")
haps = haplotype(data)
Net <- haploNet(haps)

# convert network of class HaploNet to more generalize network object 
# this is an undirected network - default is directed so you must set this option
Net.sqor <- as.network(Net, directed = FALSE)

# This conversion loosing info about edge attributes 
# We can generate this after
Net.sqor

# Get the edge list out of the network 
Net.sqor.edgelist <- as.edgelist(Net.sqor)
# convert edge list to dataframe
sqor.network<- as.data.frame(Net.sqor.edgelist)
# write table to be visualized in cytoscape
write.csv(Net.sqor.edgelist, file = "sqor.network.csv",  row.names = FALSE)
```

```{r sqor.networkor net info}
# determine distance between haplotypes (numbe of mutations) 
# this outputs a pairwise dist matrix. 
d <- dist.dna(haps, "N")
# add distance info ad edge links
nt <- rmst(d, quiet = TRUE)

# summarize the haplotypes to get totals (node sizes) and add labels 
sz <- summary(haps)
nt.labs <- attr(nt, "labels")
sz <- sz[nt.labs]

# generate table that connects each individual to their haplotypes (indv as cols, haps as rows)
ind.hap<-with(stack(setNames(attr(haps, 'index'), rownames(haps))), 
table(hap=ind, individuals=rownames(data)[values]))

# convert to a data frame
mydata <- as.data.frame(ind.hap)

# remove extra columns where there is no haplotype
good <- mydata[mydata$Freq == 1,]

# Split the indv names to remove the _H1 and _H2 that extists in labels 
indv <- strsplit(as.character(good$individuals), '_')

# split to get location by removing sample number 
locations <- strsplit(as.character(good$individuals), "(?=[A-Za-z])(?<=[0-9])|(?=[0-9])(?<=[A-Za-z])", perl=TRUE)
locations <- sapply(locations, "[[", 1)

# add location information to the table of hapltypes 
new.hap<- table(good$hap, locations)
# convert to data frame
new.hap <- as.data.frame(new.hap)
#rename the first column that contains haplotype info 
new.hap<-rename(new.hap, "Haplotype" = "Var1")

# convert roman numerals to numbers
new.hap$Haplotype <- as.numeric(new.hap$Haplotype)

# pivot the table to wide for annotations in cytoscape and save the table
new.hap.wide <- pivot_wider(new.hap, names_from = "locations", values_from = "Freq")
write.csv(new.hap.wide, "sqor.hap.info.wide.csv", quote = FALSE, row.names = FALSE)

# set step info table
# I could not convert the haploNet object to a datatable and retain step info
# As a workaround, I save the network then read it in as a data frame
write.csv(Net, "sqor.step.info.csv", quote = FALSE, row.names = FALSE)
step_size <- read.csv("sqor.step.info.csv")

# set edge interactions so cytoscape can read 
step_size$`shared name` <- paste0(step_size$X.1, " (interacts with) ", step_size$X)

# only select the interaction and step size table
step_size %>%
  select(`shared name`, step) -> step_size
# write step size table 
write.csv(step_size, "sqor.step.info.csv", quote = FALSE, row.names = FALSE)
```








```{r sqor singltons}
no_sing_haps <- subset(haps, minfreq = 2)
no_sing_haps
Net_1 <- haploNet(no_sing_haps)
d <- dist.dna(no_sing_haps, "N")
nt <- rmst(d, quiet = TRUE)
nt
sz <- summary(no_sing_haps)
nt.labs <- attr(nt, "labels")
sz <- sz[nt.labs]
#plot(Net_1, size = sz)

Net.sqor <- as.network(Net_1, directed = FALSE)
# Get the edge list out of the network 
Net.sqor.edgelist <- as.edgelist(Net.sqor)
# convert edge list to dataframe
sqor.network<- as.data.frame(Net.sqor.edgelist)
# write table to be visualized in cytoscape
write.csv(Net.sqor.edgelist, file = "sqor_rm1.network.csv",  row.names = FALSE)


ind.hap<-with(stack(setNames(attr(no_sing_haps, 'index'), rownames(no_sing_haps))), 
table(hap=ind, individuals=rownames(data)[values]))
mydata <- as.data.frame(ind.hap)
good <- mydata[mydata$Freq == 1,]
indv <- strsplit(as.character(good$individuals), '_')
locations <- strsplit(as.character(good$individuals), "(?=[A-Za-z])(?<=[0-9])|(?=[0-9])(?<=[A-Za-z])", perl=TRUE)
locations <- sapply(locations, "[[", 1)

# add location information to the table of haplotypes 
new.hap<- table(good$hap, locations)
# convert to data frame
new.hap <- as.data.frame(new.hap)
#rename the first column that contains haplotype info 
new.hap<-rename(new.hap, "Haplotype" = "Var1")
# convert roman numerals to numbers
new.hap$Haplotype <- as.numeric(new.hap$Haplotype)

# pivot the table to wide for annotations in cytoscape and save the table
new.hap.wide <- pivot_wider(new.hap, names_from = "locations", values_from = "Freq")
write.csv(new.hap.wide, "sqor_rm1.hap.info.wide.csv", quote = FALSE, row.names = FALSE)

# set step info table
# I could not convert the haploNet object to a datatable and retain step info
# As a workaround, I save the network then read it in as a data frame
write.csv(Net_1, "sqor_rm1.step.info.csv", quote = FALSE, row.names = FALSE)
step_size <- read.csv("sqor_rm1.step.info.csv")

# set edge interactions so cytoscape can read 
step_size$`shared name` <- paste0(step_size$X.1, " (interacts with) ", step_size$X)

# only select the interaction and step size table
step_size %>%
  select(`shared name`, step) -> step_size
# write step size table 
write.csv(step_size, "sqor_rm1.step.info.csv", quote = FALSE, row.names = FALSE)


new.hap

haps
```


```{r sqor singltons}
pdf("sqor_rm1.pdf")
plot(Net_1, size=attr(Net_1, "freq"), scale.ratio = 2, cex = 0.8, pie=new.hap, bg = colors, fast = FALSE)
#legend('topleft', colnames(new.hap), col=rainbow(ncol(new.hap)), bg = colors, pch=20, cex=1)
dev.off()

pdf("sqor_rm0.pdf")
plot(Net, size=attr(Net_1, "freq"), scale.ratio = 2, cex = 0.8, pie=new.hap, bg = colors, fast = FALSE)
dev.off()
```


```{r sqor doubletons removed}

no_doub_haps <- subset(haps, minfreq = 3 )
no_doub_haps 
Net_2 = haploNet(no_doub_haps)
d <- dist.dna(no_doub_haps, "N")
nt <- rmst(d, quiet = TRUE)
nt
sz <- summary(no_doub_haps)
nt.labs <- attr(nt, "labels")
sz <- sz[nt.labs]
plot(Net_2, size = sz)
ind.hap<-with(stack(setNames(attr(no_doub_haps, 'index'), rownames(no_doub_haps))), 
table(hap=ind, individuals=rownames(data)[values]))
ind.hap
mydata <- as.data.frame(ind.hap)

good <- mydata[mydata$Freq == 1,]
good
indv <- strsplit(as.character(good$individuals), '_')
locations <- strsplit(as.character(good$individuals), "(?=[A-Za-z])(?<=[0-9])|(?=[0-9])(?<=[A-Za-z])", perl=TRUE)
locations <- sapply(locations, "[[", 1)
new.hap<- table(good$hap, locations)
summary(new.hap)

colors = c('#EEE8AA', '#CCFFFF', '#FFFF33', '#FFDB58', '#98AFC7', '#EAC117', '#FBB117', '#151B54', '#1F45FC' ,'#00BFFF' )


#pdf("sqor_rm2.pdf")
plot(Net_2, size=attr(Net_2, "freq"), scale.ratio = 2, cex = 0.8, pie=new.hap, fast = FALSE, legend = c(-240,-20) )
#legend('topleft', colnames(new.hap), col=rainbow(ncol(new.hap)), pch=20, cex=1)

#dev.off()

# rep of 290 haplotypes total out of 398
290/398
# rep of 73% of the data 
```

version























## Info about SQR VCF
```{r}

var_qual <- read_delim("/Users/kararyan/Downloads/sqr_vcf_out.lqual", delim = "\t",
           col_names = c("chr", "pos", "qual"), skip = 1)

a <- ggplot(var_qual, aes(qual)) + 
  geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3) + 
  theme_light() +
  scale_x_continuous(labels = comma)
a

summary(var_qual)

var_depth <- read_delim("/Users/kararyan/Downloads/sqr_vcf_out.ldepth.mean", delim = "\t",
           col_names = c("chr", "pos", "mean_depth", "var_depth"), skip = 1)
summary(var_depth$mean_depth)

var_miss <- read_delim("/Users/kararyan/Downloads/sqr_vcf_out.lmiss", delim = "\t",
                       col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), skip = 1)
summary(var_miss)

ind_depth <- read_delim("/Users/kararyan/Downloads/sqr_vcf_out.idepth", delim = "\t",
                        col_names = c("ind", "nsites", "depth"), skip = 1)
summary(ind_depth)
a <- ggplot(ind_depth, aes(depth)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()


ind_miss  <- read_delim("/Users/kararyan/Downloads/sqr_vcf_out.imiss", delim = "\t",
                        col_names = c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(ind_miss, aes(fmiss)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3, bins = 15)
a + theme_light()
summary(ind_miss)
ind_miss %>%
  filter(nmiss > 0)
```















```{r}
#sqr <- read.vcfR("~/OneDrive - Washington State University (email.wsu.edu)/Research/Capture_Data/capture.bed.merged.geno.m90.biallelic.SNPs.Q30.DP10.maf01.nVS44.sqr.recode.vcf")


sqr<- read.FASTA("/Users/kararyan/Downloads/capture.bed.merged.geno.m90.biallelic.SNPs.Q30.DP10.maf01.nVS44.sqr.allsites.recode.min4.fasta")



sqr_haps<- haplotype(sqr)

sqr_haps_subset <- subset(sqr_haps, minfreq = 2)
sqr_haps_subset
sqr_haps
#sum(sqr_haps_sum)
d <- dist.dna(sqr_haps_subset, "N")
nt <- rmst(d, quiet = TRUE)
nt
sz <- summary(sqr_haps_subset)
nt.labs <- attr(nt, "labels")
sz <- sz[nt.labs]

sqrNet <- haploNet(nt, size = )
plot(nt, size = sz)
```

```{r}

sqr2 <- read.FASTA("/Users/kararyan/Downloads/capture.bed.merged.geno.m90.biallelic.SNPs.Q30.DP10.maf01.nVS44.sqr.recode.min100.fasta")
sqr2_haps <- haplotype(sqr2 )
sqr2_haps <-subset(sqr2_haps, minfreq = 4)
summary(sqr2_haps)


d <- dist.dna(sqr2_haps, "N")
nt <- rmst(d, quiet = TRUE)
sz <- summary(sqr2_haps)
nt.labs <- attr(nt, "labels")
nt.labs
sz <- sz[nt.labs]
nt
sqrNet <- haploNet(sqr2_haps, d = d)
sqrNet
plot(sqrNet, threshold = c(1,10))
```





















